test/request_reply_message_publisher_test.go

```diff
@@ -130,4 +134,421 @@ var _ = Describe("RequestReplyPublisher", func() {
+			messageBuilder = messagingService.MessageBuilder()
+		})
+
+		AfterEach(func() {
+			if messagingService.IsConnected() {
+				helpers.DisconnectMessagingService(messagingService)
+			}
+		})
+
+		It("can print the publisher to a string and see the pointer", func() {
+			publisher, err := messagingService.RequestReply().CreateRequestReplyMessagePublisherBuilder().Build()
+			Expect(err).ToNot(HaveOccurred())
+			str := fmt.Sprint(publisher)
+			Expect(str).ToNot(BeEmpty())
+			Expect(str).To(ContainSubstring(fmt.Sprintf("%p", publisher)))
+		})
+
+		Describe("request-reply publisher termination tests", func() {
+			publishTopic := resource.TopicOf("hello/world")
+			topicSubscription := resource.TopicSubscriptionOf("hello/world")
+			largeByteArray := make([]byte, 16384)
+			timeOut := 5 * time.Second
+
+			// A handler for the request-reply publisher
+			publisherReplyHandler := func(message message.InboundMessage, userContext interface{}, err error) {
+				if err == nil { // Good, a reply was received
+					Expect(message).ToNot(BeNil())
+					payload, _ := message.GetPayloadAsString()
+					fmt.Printf("The reply inbound payload: %s\n", payload)
+				} else if terr, ok := err.(*solace.TimeoutError); ok { // Not good, a timeout occurred and no reply was received
+					// message should be nil
+					// This handles the situation that the requester application did not receive a reply for the published message within the specified timeout.
+					// This would be a good location for implementing resiliency or retry mechanisms.
+					fmt.Printf("The message reply timed out with %s\n", terr)
+				} else { // async error occurred.
+					Expect(err).ToNot(BeNil())
+				}
+			}
+
+			// A helper function to saturate a given publisher. Counts the number of published messages at the given int pointer
+			// Returns a channel that is closed when the publisher receives an error from a call to Publish
+			// Returns a channel that can be closed when the test completes, ie. if an error occurred
+			publisherSaturation := func(publisher solace.RequestReplyMessagePublisher, publishedMessages *int) (publisherComplete, testComplete chan struct{}) {
+				publisherComplete = make(chan struct{})
+				testComplete = make(chan struct{})
+
+				toPublish, err := messageBuilder.BuildWithByteArrayPayload(largeByteArray)
+				Expect(err).ToNot(HaveOccurred())
+
+				go func() {
+					defer GinkgoRecover()
+				loop:
+					for {
+						select {
+						case <-testComplete:
+							break loop
+						default:
+						}
+						err := publisher.Publish(toPublish, publisherReplyHandler, publishTopic, timeOut, nil /* properties */, nil /* usercontext */)
+
+						if err != nil {
+							Expect(err).To(BeAssignableToTypeOf(&solace.IllegalStateError{}))
+							break loop
+						} else {
+							(*publishedMessages)++
+						}
+					}
+					close(publisherComplete)
+				}()
+				// allow the goroutine above to saturate the publisher
+				time.Sleep(100 * time.Millisecond)
```

Comment:
You sleep in the saturation function. And also outside of this. Typically if you got a sleep you just need a channel with a select statement instead. Why do need this sleep and what ways can it be eliminated?

Reply:
Truth is we don't. I will get rid of the time.Sleep in the saturation function and use select in places where we need timing control.